File: lima_linguisticdata/scripts/buildSeparableVerbList.pl
#!/usr/bin/perl -s

# Copyright 2002-2013 CEA LIST
# SPDX-FileCopyrightText: 2022 CEA LIST <gael.de-chalendar@cea.fr>
#
# SPDX-License-Identifier: MIT

# Help mode
if ($main::h || $main::help) {
    print <<EOF;
extract a list of reflexive verbs from the list of idiomatic
expressions and store them in a file as a class used for
automaton rules
usage buildReflexiveVerbList.pl idiom-file
EOF
    exit;
}

use strict;

die "need output file for lists (-list=..)" unless $main::list;
die "need output file for rules (-rules=..)" unless $main::rules;

open(FLIST,">$main::list") 
    || die "cannot open output file \"$main::list\"";

open(FRULES,">$main::rules") 
    || die "cannot open output file \"$main::rules\"";

my (%ReflexiveVerbs);

while (<>) {
    if (/;verbe pronominal;/) {
        if (/^[^;]*;[^;]*;([^;]*);([^;]*);([^;]*);/) {
            my ($abs,$trigger,$exp)=($1,$2,$3); 
            $abs='C' if ($abs eq ""); # C(contextual) if not A(bsolute)
            if ($trigger =~ /\[V\]/) {
                if ($exp eq "s' (pp) $trigger" ||
                    $exp eq "se (pp) $trigger") {
                    $trigger =~ s/\[V\]\&?//;
                    push @{$ReflexiveVerbs{$abs}}, $trigger;
                }
                else { # rule contain more than juste se+verb
                    &addReflexiveVerbRule($abs,$trigger,$exp,1);
                }
            }
            else { # verb is not trigger
                if ($exp=~/\[V\]\&?([^ ]*) /) {
                    &addReflexiveVerbRule($abs,$trigger,$exp,0);
                }
                else { # keep rule as is
                    print STDERR "Warning: did not know how to deal ".
                        "with pronominal expression $exp\n";
                    print;
                }
            }
        }
        else {
            print;
        }
    }
#     elsif (/\(.*\)/ && /;verbe/) {
#         # separable verb expressions (prendre part)
#     }
    else {
        print;
    }
}

if (exists $ReflexiveVerbs{'C'}) {
    print FLIST "\@ReflexiveVerbs=(".
        join("\$V,\n",@{$ReflexiveVerbs{'C'}})."\$V)\n\n";
}
else {
    print STDERR "Warning: no contextual rule for pronominal verb\n";
}
if (exists $ReflexiveVerbs{'A'}) {
    print FLIST "\@AbsReflexiveVerbs=(".
        join("\$V,\n",@{$ReflexiveVerbs{'A'}})."\$V)\n\n";
}
else {
    print STDERR "Warning: no absolute rule for pronominal verb\n";
}

close(FLIST);
close(FRULES);

# add a verb rule for expression that is more complex than
# just reflexive pronoun + verb (s'en aller, s'emmeler les pinceaux)
# rule can be trigger on something else than the verb (pinceaux in
# s'emmeler les pinceaux)
sub addReflexiveVerbRule {
    my ($abs,$trigger,$exp,$triggerIsVerb) =@_;
    my ($pron,$pre,$verb,$post);
    if ($exp =~ /(s\'|se)(.*) (\[V\]\&?[^ ]*)(.*)/) {
        ($pron,$pre,$verb,$post)=($1,$2,$3,$4);
    }
    # remove possible space at beginning of pre and post
    $pre=~s/^ *//;
    $post=~s/^ *//;
    
    #print STDERR "--[pron=$pron|pre=$pre|verb=$verb|post=$post]--\n";
    
    $pre=~s/\(pp\)//; # remove (pp) => rewritten in rule
    $post=~s/\(D\)/[(\$ADV?)]/g; 

    $pre=~s/^ *//;
    $post=~s/^ *//;

    if ($triggerIsVerb) {
        $trigger =~ s/\[V\]//;
        $trigger.='$V';

        # cannot determine if possible pronouns,auxiliaries and adverbs
        # are before or after $pre (s'en aller => je m'en suis allé,
        # s'entre-déchirer => je me suis entre-déchiré)
        # => set both possibilities
        print FRULES
            "$trigger".
            ':$PRON-CLR';
        if ($pre ne "") {
            print FRULES "(($pre ".
            '[$PRON?] [(@AuxiliaireEtre ($ADV){0-2})?])'.
            '|([$PRON?] [(@AuxiliaireEtre ($ADV){0-2})?] '.
            "$pre))";
        }
        else {
            print FRULES 
                '([$PRON?] [(@AuxiliaireEtre ($ADV){0-2})?])';
        }
        print FRULES ':'.
            "$post".
            ':ABS_IDIOM$V---p---:'."\n".
            '+AgreementConstraint(trigger.1,left.1,"NUMBER")'."\n".
            '+AgreementConstraint(trigger.1,left.1,"PERSON")'."\n";
    }
    else {
        $verb =~ s/\[V\]//;
        $verb.='$V';

        # find trigger in rule, either in pre or post
        my $quotemetaTrigger=quotemeta($trigger);
        my ($left,$right);
        if ($pre =~ /(.*)(^| )$quotemetaTrigger( |$)(.*)/)  {
            $left=$1; 
            $right=$4;
            print FRULES
                "$trigger".
                ':$PRON-CLR';
            print FRULES " $left" if $left;
            print FRULES ':';
            print FRULES "$right " if $right;
            print FRULES '[$PRON?] [(@AuxiliaireEtre ($ADV){0-2})?]'.
                " $verb".
                ':ABS_IDIOM$V---p---:'."\n".
                '+AgreementConstraint(left.1,right.4,"NUMBER")'."\n".
                '+AgreementConstraint(left.1,right.4,"PERSON")'."\n";
        }
        elsif ($post =~ /(.*)(^| )$quotemetaTrigger( |$)(.*)/)  {
            $left=$1; 
            $right=$4;
            print FRULES
                "$trigger".
                ':$PRON-CLR ';
            if ($pre ne "") {
                print FRULES "(($pre ".
                    '[$PRON?] [(@AuxiliaireEtre ($ADV){0-2})?])'.
                    '|([$PRON?] [(@AuxiliaireEtre ($ADV){0-2})?] '.
                    "$pre))";
            }
            else {
                print FRULES 
                    '([$PRON?] [(@AuxiliaireEtre ($ADV){0-2})?])';
            }
            print FRULES " $verb";
            print FRULES " $left" if $left;
            print FRULES ':';
            print FRULES "$right " if $right;
            print FRULES ':ABS_IDIOM$V---p---:'."\n".
                '+AgreementConstraint(left.1,left.3,"NUMBER")'."\n".
                '+AgreementConstraint(left.1,left.3,"PERSON")'."\n";
        }
        else {
            print STDERR "Warning : cannot find trigger [".$trigger.
                "] in expression \"$exp\"\n";
        }
    }
}

File: lima_linguisticdata/scripts/build_separable_verb_list.py
import argparse
import re
import sys
from collections import defaultdict

from tqdm import tqdm


def add_reflexive_verb_rule(abs_type, trigger, exp, trigger_is_verb,
                            rules_file):
    match = re.match(r"(s'|se)(.*) (\[V\]\&?[^ ]*)(.*)", exp)
    if match:
        pron, pre, verb, post = match.groups()
        pre = pre.strip()
        post = post.strip()

        pre = re.sub(r'\(pp\)', '', pre)
        post = re.sub(r'\(D\)', '[($ADV?)]', post)

        with open(rules_file, 'a') as frules:
            if trigger_is_verb:
                trigger = re.sub(r'\[V\]', '', trigger) + '$V'
                frules.write(f"{trigger}:$PRON-CLR")
                if pre:
                    frules.write(
                        f"(({pre} [$PRON?] [@AuxiliaireEtre ($ADV){{0-2}}])"
                        f"|([$PRON?] [(@AuxiliaireEtre ($ADV){{0-2}})?] "
                        f"{pre}))")
                else:
                    frules.write('([$PRON?] [@AuxiliaireEtre ($ADV){0-2})?])')
                frules.write(f":{post}:ABS_IDIOM$V---p---:\n")
                frules.write(
                    '+AgreementConstraint(trigger.1,left.1,"NUMBER")\n')
                frules.write(
                    '+AgreementConstraint(trigger.1,left.1,"PERSON")\n')
            else:
                verb = re.sub(r'\[V\]', '', verb) + '$V'
                quotemeta_trigger = re.escape(trigger)
                left, right = '', ''

                if re.search(f"(.*)(^| ){quotemeta_trigger}( |$)(.*)", pre):
                    left, right = re.split(
                        f"(.*)(^| ){quotemeta_trigger}( |$)(.*)", pre)[0::3]
                elif re.search(f"(.*)(^| ){quotemeta_trigger}( |$)(.*)", post):
                    left, right = re.split(
                        f"(.*)(^| ){quotemeta_trigger}( |$)(.*)", post)[0::3]

                if left or right:
                    frules.write(f"{trigger}:$PRON-CLR {left if left else ''}:"
                                 f"{right if right else ''} [$PRON?] "
                                 f"[(@AuxiliaireEtre ($ADV){{0-2}})] {verb}:"
                                 f"{abs_type.upper()}_IDIOM$V---p---:\n")
                    frules.write(f'+AgreementConstraint('
                                 f'left.1,right.{4 if left else 3},'
                                 f'"NUMBER")\n')
                    frules.write(f'+AgreementConstraint('
                                 f'left.1,right.{4 if left else 3},'
                                 f'"PERSON")\n')
                else:
                    sys.stderr.write(f"Warning: cannot find trigger "
                                     f"[{trigger}] in expression \"{exp}\"\n")
    else:
        sys.stderr.write(f"Warning: failed to parse complex reflexive verb "
                         f"rule in expression \"{exp}\"\n")


def main():
    parser = argparse.ArgumentParser(
        description='Extract a list of reflexive verbs from the list of '
                    'idiomatic expressions and store them in a file as a '
                    'class used for automaton rules.')
    parser.add_argument('idiom_file', type=argparse.FileType('r'),
                        help='Input file containing idiomatic expressions')
    parser.add_argument('-list', required=True, help='Output file for lists')
    parser.add_argument('-rules', required=True, help='Output file for rules')

    args = parser.parse_args()

    reflexive_verbs = defaultdict(list)
    idiom_lines = args.idiom_file.readlines()

    with tqdm(total=len(idiom_lines),
              desc="Processing idiomatic expressions") as pbar:
        for line in idiom_lines:
            pbar.update(1)
            if ';verbe pronominal;' in line:
                match = re.match(
                    r'^[^;]*;[^;]*;([^;]*);([^;]*);([^;]*);', line)
                if match:
                    abs_type, trigger, exp = match.groups()
                    abs_type = 'C' if abs_type == '' else abs_type
                    if '[V]' in trigger:
                        if exp in [f"s' (pp) {trigger}", f"se (pp) {trigger}"]:
                            trigger = re.sub(r'\[V\]\&?', '', trigger)
                            reflexive_verbs[abs_type].append(trigger)
                        else:
                            add_reflexive_verb_rule(abs_type, trigger, exp,
                                                    True, args.rules)
                    else:
                        if re.search(r'\[V\]\&?([^ ]*) ', exp):
                            add_reflexive_verb_rule(abs_type, trigger, exp,
                                                    False, args.rules)
                        else:
                            sys.stderr.write(
                                f"Warning: did not know how to deal with "
                                f"pronominal expression {exp}\n")
                            sys.stdout.write(line)
                else:
                    sys.stdout.write(line)
            else:
                sys.stdout.write(line)

    with open(args.list, 'w') as flist:
        if 'C' in reflexive_verbs:
            rv = ',\n'.join(f'{verb}$V' for verb in reflexive_verbs['C'])
            flist.write(f"@ReflexiveVerbs=({rv}$V)\n\n")
        else:
            sys.stderr.write(
                "Warning: no contextual rule for pronominal verb\n")

        if 'A' in reflexive_verbs:
            rv = ',\n'.join(f'{verb}$V' for verb in reflexive_verbs['A'])
            flist.write(f"@AbsReflexiveVerbs=({rv}$V)\n\n")
        else:
            sys.stderr.write("Warning: no absolute rule for pronominal verb\n")


if __name__ == "__main__":
    main()
